wasm/jit/jit_arm64.go

```diff
@@ -862,19 +862,113 @@ func (c *arm64Compiler) compileXor(o *wazeroir.OperationXor) error {
 }
 
 func (c *arm64Compiler) compileShl(o *wazeroir.OperationShl) error {
-	return fmt.Errorf("TODO: unsupported on arm64")
+	x1, x2, err := c.popTwoValuesOnRegisters()
+	if err != nil {
+		return err
+	}
+
+	if isZeroRegister(x1.register) || isZeroRegister(x2.register) {
+		c.locationStack.pushValueLocationOnRegister(x1.register)
+		return nil
+	}
+
+	var inst obj.As
+	switch o.Type {
+	case wazeroir.UnsignedInt32:
+		inst = arm64.ALSLW
+	case wazeroir.UnsignedInt64:
+		inst = arm64.ALSL
+	}
+
+	c.applyTwoRegistersToRegisterInstruction(inst, x2.register, x1.register, x1.register)
+	c.locationStack.pushValueLocationOnRegister(x1.register)
+	return nil
 }
 
 func (c *arm64Compiler) compileShr(o *wazeroir.OperationShr) error {
-	return fmt.Errorf("TODO: unsupported on arm64")
+	x1, x2, err := c.popTwoValuesOnRegisters()
+	if err != nil {
+		return err
+	}
+
+	if isZeroRegister(x1.register) || isZeroRegister(x2.register) {
+		c.locationStack.pushValueLocationOnRegister(x1.register)
+		return nil
+	}
+
+	var inst obj.As
+	switch o.Type {
+	case wazeroir.SignedInt32:
+		inst = arm64.AASRW
+	case wazeroir.SignedInt64:
+		inst = arm64.AASR
+	case wazeroir.SignedUint32:
+		inst = arm64.ALSRW
+	case wazeroir.SignedUint64:
+		inst = arm64.ALSR
+	}
+
+	c.applyTwoRegistersToRegisterInstruction(inst, x2.register, x1.register, x1.register)
+	c.locationStack.pushValueLocationOnRegister(x1.register)
+	return nil
 }
 
 func (c *arm64Compiler) compileRotl(o *wazeroir.OperationRotl) error {
-	return fmt.Errorf("TODO: unsupported on arm64")
+	x1, x2, err := c.popTwoValuesOnRegisters()
+	if err != nil {
+		return err
+	}
+
+	if isZeroRegister(x1.register) || isZeroRegister(x2.register) {
+		c.locationStack.pushValueLocationOnRegister(x1.register)
+		return nil
+	}
+
+	var (
+		inst    obj.As
+		neginst obj.As
+	)
+
+	switch o.Type {
+	case wazeroir.UnsignedInt32:
+		inst = arm64.ARORW
+		neginst = arm64.ANEGW
+	case wazeroir.UnsignedInt64:
+		inst = arm64.AROR
+		neginst = arm64.ANEG
+	}
+
+	// Arm64 doesn't have rotate left instruction.
+	// The shift amount needs to be converted to a negative number.
+	c.applyRegisterToRegisterInstruction(neginst, x2.register, x2.register)
```

Comment:
I adopted this way because I checked Go Assembly output of the `bits.RotateLeft()` [sample code](https://pkg.go.dev/math/bits#RotateLeft32) described in the official documentation, and it looked like this:

```
...
0x0258 00600 (main.go:11)       MOVD    $15, R0
0x025c 00604 (main.go:11)       MOVD    $2, R1
0x0260 00608 (main.go:11)       RORW    R1, R0, R0
...
0x0130 00304 (main.go:10)       MOVD    $15, R0
0x0134 00308 (main.go:10)       MOVD    $-2, R1
0x0138 00312 (main.go:10)       RORW    R1, R0, R1
...
```

I think this is valid, but if there is a better way, I'd love to hear your comments!

Reply:
ok confirmed that NEG is actually used in register allocated case

```go
func fooooo(x uint32, k int) uint32 {
	return bits.RotateLeft32(x, k)
}
```

is compiled to 

```
"".fooooo STEXT size=32 args=0x18 locals=0x0 funcid=0x0 leaf
	0x0000 00000 (main.go:12)	TEXT	"".fooooo(SB), LEAF|NOFRAME|ABIInternal, $0-24
	0x0000 00000 (main.go:12)	FUNCDATA	ZR, gclocals·33cdeccccebe80329f1fdbee7f5874cb(SB)
	0x0000 00000 (main.go:12)	FUNCDATA	$1, gclocals·33cdeccccebe80329f1fdbee7f5874cb(SB)
	0x0000 00000 (main.go:12)	FUNCDATA	$5, "".fooooo.arginfo1(SB)
	0x0000 00000 (main.go:13)	MOVD	"".k+8(FP), R0
	0x0004 00004 (main.go:13)	NEG	R0, R0
	0x0008 00008 (main.go:13)	MOVWU	"".x(FP), R1
	0x000c 00012 (main.go:13)	RORW	R0, R1, R0
	0x0010 00016 (main.go:13)	MOVW	R0, "".~r2+16(FP)
	0x0014 00020 (main.go:13)	RET	(R30)
	0x0000 e0 0b 40 f9 e0 03 00 cb e1 0b 40 b9 20 2c c0 1a  ..@.......@. ,..
	0x0010 e0 1b 00 b9 c0 03 5f d6 00 00 00 00 00 00 00 00  ......_.........
```

Thank you for your efforts!